### ISSUE TITLE
[feature][byob] Support more secrets

### ISSUE THREAD

**Issue by ianlewis** [slsa-framework/slsa-github-generator#1790]
BYOB supports secrets but has a fixed number of secrets it can support. We should support more than the current two secrets.

**Comment by laurentsimon** [comment]
What use case did you encounter for a 3rd secret?

**Comment by ianlewis** [comment]
> What use case did you encounter for a 3rd secret?

So far I haven't encountered a situation where a 3rd secret would be needed, so I think it's fairly low priority but I figure we should support more as adding more secret inputs later would be a breaking change. For example, a user's internal action would fail if the delegator passed `slsa-workflow-secret3` but it wasn't defined in their action.

**Comment by laurentsimon** [comment]
Looks like we need more secrets then! We can add an arbitrary number of secrets. Today we have only 2 secrets supported but it's easy to add more.

> inheritance

Secrets may be inherited only if the re-usable workflow is in the same organization

> The question is, can a PW set secrets in such a way that they get inherited all the way through the chain til they reach the callback action?

Yes. Right now we do that like in https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/nodejs-action/action.yml#L15-L23. The TRW secrets are defined in https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/delegator_generic_slsa3.yml#L33-L44. So the TRW can declare any secrets, and we forward them to the Action, which knows the mapping.

Will that work for your use case?

**Comment by aalmiray** [comment]
Ah OK. So I suppose by convention a given TRW and its companion callback could map `secret1` as the value for `GPG_PASSPHRASE`. It's the callback's responsibility to route secret1 to where it needs to be. 

How many secrets would be needed? hard to tell. Here's an example kitchen-sink repository with more than 40 secrets https://github.com/jreleaser/smoketests-jreleaser/blob/main/.jreleaser/config.toml

JReleaser's official build currently relies on 17 secrets https://github.com/jreleaser/jreleaser/blob/ceea4122057c856767dc3c3d3089679433dfe5dc/.github/workflows/release.yml#L131

**Comment by laurentsimon** [comment]
> Ah OK. So I suppose by convention a given TRW and its companion callback could map secret1 as the value for GPG_PASSPHRASE. It's the callback's responsibility to route secret1 to where it needs to be.

Correct, at least in our design. If you think of something better, let us know.

> How many secrets would be needed? hard to tell. Here's an example kitchen-sink repository with more than 40 secrets https://github.com/jreleaser/smoketests-jreleaser/blob/main/.jreleaser/config.toml

That's a lot indeed! Note that the majority of the secrets are not actual secrets (username have low entropy, so do emails, etc). We recommend passing them as standard TRW inputs instead and have the TRW use the masked inputs https://github.com/slsa-framework/slsa-github-generator/blob/main/actions/delegator/setup-token/action.yml#L50-L60 to prevent having them appear in the final provenance. Secrets, passwords and tokens should be passed a secrets.

One of the reason for this ie because GitHub has some lose scanning that sometimes gets triggered as false positive https://github.com/slsa-framework/slsa-github-generator/issues/1236

**Comment by ianlewis** [comment]
> Looks like we need more secrets then! We can add an arbitrary number of secrets. Today we have only 2 secrets supported but it's easy to add more.

This is a bit tricky since we have to add them as inputs to the delegator workflow and the delegator workflow passes them to the internal action. So if a TRW upgrades the delegator workflow it may start breaking things if the internal action isn't also updated to support the new secrets.

**Comment by laurentsimon** [comment]
We can version the workflow to bump the major version though?

We need to verify whether not providing a required input actual breaks things, or jut puts up a warning in the logs. I recall see warning in the logs but not necessarily failures.

**Comment by ianlewis** [comment]
yeah, I believe we'll have to bump it to mark the backwards incompatibility. IIRC I tested this and it fails when the inputs are not defined.

**Comment by laurentsimon** [comment]
Just tested this and it seems to jut show `Warning: Unexpected input(s) 'not-exist', valid inputs are ['xxx']
`

**Comment by ianlewis** [comment]
> Just tested this and it seems to just show `Warning: Unexpected input(s) 'not-exist', valid inputs are ['xxx'] `

Hmm, I stand corrected then I guess. I remembered differently.

**Comment by aalmiray** [comment]
@laurentsimon I'm afraid defining input fields to be masked is not enough unless the PW passes masked inputs from the start

Invoking a RW with 3 inputs and a few secrets https://github.com/kordamp/kordamp-gradle-plugins/blob/master/.github/workflows/release.yml

GH prints out the RW's inputs at line 39
https://github.com/kordamp/kordamp-gradle-plugins/actions/runs/4779177846/jobs/8496085719

**Comment by ianlewis** [comment]
Yeah, IIRC the masked inputs are there to mask them from the provenance, not necessarily to mask them from the GitHub logs. They're mostly useful for values like `username`, `email` etc. that are (semi)"private" rather than "secret" values. Unfortunately they'll still be there in the logs for public repos but at least that data expires at some point and you have some access controls for it. You kinda have this issue with normal actions as well anyway.

The really annoying thing is that if you put things like username, email, or base64 values in secrets, GitHub will mask values that match even **part** of the value in **any** secret. This results in tons of normal input values getting dropped and failing builds so you're kind of forced to put them in normal inputs.

**Comment by aalmiray** [comment]
> Yeah, IIRC the masked inputs are there to mask them from the provenance, not necessarily to mask them from the GitHub logs. 

Ah. OK. I misunderstood the use of masked inputs in this case. Thank you for clarifying it 😅 

> They're mostly useful for values like `username`, `email` etc. that are (semi)"private" rather than "secret" values. Unfortunately they'll still be there in the logs for public repos but at least that data expires at some point and you have some access controls for it. You kinda have this issue with normal actions as well anyway.

Well, sensitive data should remain private if the owner deems it to be that way.

**Comment by ianlewis** [comment]
> Well, sensitive data should remain private if the owner deems it to be that way.

Right. At the very least users can make their logs inaccessible to the public via GitHub access controls, and the masked inputs will keep them from being included in the provenance.

I wish there were better options to mask secret/private inputs from logs without breaking builds, but we unfortunately don't have much control over how GitHub logs and drops/masks inputs.

**Comment by laurentsimon** [comment]
Since this issue is coming up again and again https://github.com/orgs/community/discussions/37942
/cc @josepalafox, @feelepxyz @bdehamer @kommendorkapten in case you can help with it
